### PR TITLE
Add support for TDK Lambda Genesys power supplies

### PR DESCRIPTION
**Pull request by mcdo0486**
This adds supports for the TDK Lambda Genesys 80-65 DC power supply and TDK Lambda Genesys 40-38 DC power supply.

This is implemented with a shared base class `pymeasure.instruments.tdk.tdk_base.TDK_Lambda_Base`. The only difference between the two models is their voltage and current ranges, which are both dynamic properties for the two instrument classes.

Tests and documentation are added as well.

### CODE REVIEW

File: pymeasure/adapters/prologix.py [hunk 1]
```diff
@@ -200,7 +200,7 @@ def ask(self, command):
     def write(self, command, **kwargs):
         """Write a string command to the instrument appending `write_termination`.
 
-        If the GPIB address in :attr:`.address` is defined, it is sent first.
+        If the GPIB address in :attr:`self.address` is defined, it is sent first.
```
Comment (bmoneke) [inline pymeasure/adapters/prologix.py:203]: Why this change?

The "attr" indicates an attribute of the class/instance. Maybe the period is superfluous.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+    def __init__(self, adapter, name="TDK-Lambda Base", address=6, **kwargs):
+        """
+        The initialization of a TDK instrument requires the current address
+        of the TDK power supply. The default address for the TDK Lambda is 6.
+
+        :param adapter: VISAAdapter instance
+        :param name: Instrument name
+        :param address: Serial port daisy chain number. Default is 6.
+        """
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:73]: The `__init__` should not have a docstring. That informations should be part of the class docstring instead.
That way, initializing the class gives the needed information regarding the parameters.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        "OK" from the read buffer. If this is not done, the adapter read buffer
+        will hold the "OK" reply until the next read command is given.
+
+        :param command: Command string to be sent to the instrument.
+        """
+
+        super().write(command)
+        if '?' not in command:
+            # clear "OK" from the adapter read buffer
+            self.read()
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:107]: Is it always an "OK" or could it be an error as well?

File: pymeasure/instruments/tdk/tdk_base.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+        confirmation that the command was received.
+
+        The default, the ``Instrument.write()`` method is not set up to
+        automatically strip out instrument confirmations. If the command
+        string does not contain a "?", then ``self.read()`` once to clear the
+        "OK" from the read buffer. If this is not done, the adapter read buffer
+        will hold the "OK" reply until the next read command is given.
+
+        :param command: Command string to be sent to the instrument.
+        """
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:102]: In the other cases with confirmations, we use check_errors to read after writing, for that reason I made #883 .

File: pymeasure/instruments/tdk/tdk_base.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+        """Set the address of the power supply.
+
+        Valid values are integers between 0 - 30 (inclusive).""",
+        validator=strict_discrete_set,
+        values=range(0, 31)
+    )
+
+    remote = Instrument.control(
+        "RMT?", "RMT %s",
+        """Get the current remote operation of the power supply.
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:124]: It is a control, so please use "Control" as the first word!

File: pymeasure/instruments/tdk/tdk_base.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+    multidrop_capability = Instrument.measurement(
+        "MDAV?",
+        """Get the multi-drop option is available on the power supply.
+
+        If return value is 0, the option is not available, if 1 it is available.
+
+        Property is UNTESTED.
+        """
+    )
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:141]: why don't you turn it into a boolean and rename it?
`multidrop_capability_available == True` is self evident, for example.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+        dynamic=True
+    )
+
+    actual_current = Instrument.measurement(
+        "MC?",
+        """Measure the actual output current.
+
+        Returns a float with five digits of precision.
+        """
+    )
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:220]: We have a naming scheme, see https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#power-supplies

So current/voltage indicates the actual value, while current_setpoint indicates the setpoint.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+
+    source_output = Instrument.control(
+        "OUT?", "OUT %s",
+        """Control the output of the power supply.
+
+        Valid values are ``'ON'`` and ``'OFF'``.
+        """,
+        validator=strict_discrete_set,
+        values=["ON", "OFF"]
+    )
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:272]: Why don't you turn it into boolean?

Regarding the naming, please have a look into https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#your-instrument-s-user-interface

So output_enabled would be a better name.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+    foldback = Instrument.control(
+        "FLD?", "FLD %s",
+        """Control the fold back protection of the power supply.
+
+        Valid values are ``'ON'`` to arm the fold back protection and ``'OFF'``
+        to cancel the fold back protection.
+        """,
+        validator=strict_discrete_set,
+        values=["ON", "OFF"]
+    )
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:283]: see above

File: pymeasure/instruments/tdk/tdk_base.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+
+    auto_restart = Instrument.control(
+        "AST?", "AST %s",
+        """Control the auto restart mode.
+
+        Valid values are ``'ON'`` and ``'OFF'``.
+        """,
+        validator=strict_discrete_set,
+        values=["ON", "OFF"]
+    )
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:326]: See above

File: pymeasure/instruments/tdk/tdk_base.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+    def shutdown(self):
+        """Safety shutdown the power supply.
+
+        Ramps the power supply down to zero current using the
+        ``self.ramp_to_current(0.0)`` method and turns the output off.
+        """
+        log.info("Shutting down %s." % self.name)
+        self.ramp_to_current(0.0)
+        self.source_output = "OFF"
+        log.info("%s has been shut down." % self.name)
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:386]: I think, that this is already done in the `Instrument`'s `shutdown` method.

File: pymeasure/instruments/tdk/tdk_gen40_38.py
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_gen40_38.py:1]: As these files are pretty small, you could put these two instruments together with the base class in a single file, if you wish so.

**Review by bmoneke — COMMENTED** [review]
Thanks for your contribution, I left a few comments regarding style.

File: pymeasure/adapters/prologix.py (hunk 1 above)
Comment (mcdo0486) [reply pymeasure/adapters/prologix.py]: This change was triggered by the following sphinx error in the first CI run:
`/pymeasure/pymeasure/pymeasure/adapters/prologix.py:docstring of pymeasure.adapters.prologix.PrologixAdapter.write:3: WARNING: more than one target found for cross-reference 'address': pymeasure.instruments.anaheimautomation.DPSeriesMotorController.address, pymeasure.instruments.tdk.tdk_gen40_38.TDK_Gen40_38.address, pymeasure.instruments.tdk.tdk_gen80_65.TDK_Gen80_65.address`

I don't know why it was triggered now since the `DPSeriesMotorController` has been around for a while. Removing the period is best, thanks for the suggestion.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 3 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: It could be an error for illegal commands and illegal programming parameters, e.g. "C01" for illegal command, but only legal commands are in the driver with strict validators. This is related to `check_errors` comment?

File: pymeasure/instruments/tdk/tdk_base.py (hunk 4 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py:91]: I see what you mean, I thought of `check_errors` only for SCPI. I'll look into using that pattern, we would only `check_errors` for non-query commands.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 5 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit

File: pymeasure/instruments/tdk/tdk_base.py (hunk 8 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit. I moved to boolean, thanks for suggestion.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 9 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit

File: pymeasure/instruments/tdk/tdk_base.py (hunk 10 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit

File: pymeasure/instruments/tdk/tdk_base.py (hunk 11 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit

File: pymeasure/instruments/tdk/tdk_gen40_38.py
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_gen40_38.py:1]: I think keeping the instruments in separate files, even if small, is more obvious for people to see at a glance what is available from the file structure. I see the latest commit for Anritsu instruments did combine a lot of instruments into one file. I'd like to keep instrument files separate, but don't feel super strongly about it.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 4 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Ok, I have implemented and tested that, moved the write override logic to `check_errors` and utilized `check_set_errors` flag.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 3 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Moved this logic to check_errors.

File: pymeasure/instruments/tdk/tdk_gen40_38.py
Comment (bmoneke) [reply pymeasure/instruments/tdk/tdk_gen40_38.py:1]: Ok, that's fine. Therefore, I wrote "if you wish so".

Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_gen40_38.py:1]: Thanks, it was a good opportunity to explain my rationale.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 6 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: I didn't see this hidden comment. This is now cast to bool.

capability and availability are adjectives with similar meaning so I left it at capability.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 7 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated those properties to setpoint in last commit.

File: tests/instruments/tdk/test_tdk_base.py [hunk 12]
```diff
@@ -50,7 +50,8 @@ def test_multidrop_capability():
             [(b"ADR 6", b"OK"),
              (b"MDAV?", b'1')]
     ) as instr:
-        assert instr.multidrop_capability == True
+        value = instr.multidrop_capability
+        assert isinstance(value, bool) and value
```
Comment (bmoneke) [inline tests/instruments/tdk/test_tdk_base.py:54]:
```suggestion
        assert instr.multidrop_capability is True
```
If you want to make a comparison with a boolean (or with None), use `is`. Your test is overly complicated.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 13]
```diff
@@ -134,10 +117,11 @@ def write(self, command):
         "MDAV?",
         """Get the multi-drop option is available on the power supply.
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:118]:
```suggestion
        """Get whether the multi-drop option is available on the power supply.
```

File: pymeasure/instruments/tdk/tdk_base.py [hunk 14]
```diff
@@ -257,29 +243,34 @@ def write(self, command):
...
+        Valid values are ``True`` and ``False``.
         """,
+        check_set_errors=True,
         validator=strict_discrete_set,
-        values=["ON", "OFF"]
+        values={True: "ON", False: "OFF"},
+        map_values=True
     )
 
     foldback = Instrument.control(
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:263]: if you rename it to
```suggestion
    foldback_enabled = Instrument.control(
```
the values True/False are self evident.

File: pymeasure/instruments/tdk/tdk_base.py [hunk 15]
```diff
@@ -257,29 +243,34 @@ def write(self, command):
...
         values=[18, 23, 46]
     )
 
-    source_output = Instrument.control(
+    output_enabled = Instrument.control(
         "OUT?", "OUT %s",
         """Control the output of the power supply.
 
-        Valid values are ``'ON'`` and ``'OFF'``.
+        Valid values are ``True`` and ``False``.
```
Comment (bmoneke) [inline pymeasure/instruments/tdk/tdk_base.py:255]: The, currently recommended, way to describe a property are described here: https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings

If you state, that the property is a boolean (e.g. by appending `(bool)` to the introduction sentence), it follows automatically, that the valid values are True or False.

**Review by bmoneke — APPROVED** [review]
The code looks good.
A few small comments.

File: pymeasure/instruments/tdk/tdk_base.py (hunk 13 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit

File: pymeasure/instruments/tdk/tdk_base.py (hunk 14 above)
Comment (mcdo0486) [reply pymeasure/instruments/tdk/tdk_base.py]: Updated in last commit

**Comment by bmoneke** [comment]
I did not follow up, after restarting the checks. Is this PR mergeable from your side?

**Comment by mcdo0486** [comment]
> I did not follow up, after restarting the checks. Is this PR mergeable from your side?

Yea it's good to go.
